### DISCUSSION TITLE
Questions and problem installing TornadoVM on Windows

### DISCUSSION THREAD

**Discussion by Excelnabb** [beehive-lab/TornadoVM#215] [Q&A]
Hi

First of all, thanks for developing such a program!

I'm trying to follow this guide to run it on Windows 10:
https://github.com/beehive-lab/TornadoVM/blob/master/tornado-assembly/src/docs/20_INSTALL_WINDOWS_WITH_GRAALVM.md

However I stumble on some problems:

1. It says that I need to install and download GraalVM Community Edition 22.2.0 (https://github.com/graalvm/graalvm-ce-builds/releases/tag/vm-22.2.0).
For the program I want to run with the GPU I'm already running with the enterprise version of the code, which is 20% faster  _graalvm-ee-java17-windows-amd64-22.3.0_ will this work? Or do I need to use the community edition 22.2.0?

2. When it says # UPDATE PATH TO ACTUAL LOCATION OF THE JDK OR GRAAL (REMEMBER OF UNIX_STYLE SLASHES AND SPACES!!!)
export JAVA_HOME="C:/graalvm-ce-java11-22.2.0" 
Does this mean the location of the Java code for the particular program I want to run with GPU? In my case: C:\StrategyQuantX\j64
(folder where I extracted the _graalvm-ee-java17-windows-amd64-22.3.0_ )

3. Lastly, when I try to compile the program, I get these errors, not sure how to fix them: [Error on building Tornado.txt](https://github.com/beehive-lab/TornadoVM/files/9978042/Error.on.building.Tornado.txt)

Thank you!

**Comment by jjfumero** [comment]
Hi @Excelnabb , 
  thanks for trying TornadoVM. Our answers below

> Hi
> 
> First of all, thanks for developing such a program!

Thank you! 

> 
> I'm trying to follow this guide to run it on Windows 10: https://github.com/beehive-lab/TornadoVM/blob/master/tornado-assembly/src/docs/20_INSTALL_WINDOWS_WITH_GRAALVM.md
> 
> However I stumble on some problems:
> 
> 1. It says that I need to install and download GraalVM Community Edition 22.2.0 (https://github.com/graalvm/graalvm-ce-builds/releases/tag/vm-22.2.0).
>    For the program I want to run with the GPU I'm already running with the enterprise version of the code, which is 20% faster  _graalvm-ee-java17-windows-amd64-22.3.0_ will this work? Or do I need to use the community edition 22.2.0?

To be honest, we haven't tried with the EE of GraamVM. Ideally, it should work. However, TornadoVM depends on the Graal IR (intermediate compiler representation). Since we haven't used the EE, we can't tell the IR scheduling and structure of the Graal IR EE is that different that TornadoVM IR cannot understand. For example, auto-vectorization. TornadoVM also does vectorization for OpenCL and SPIR-V using the Graal IR in a specific format. What we know is that it should work well with Graal CE. But feel free to give it a try.


> 2. When it says # UPDATE PATH TO ACTUAL LOCATION OF THE JDK OR GRAAL (REMEMBER OF UNIX_STYLE SLASHES AND SPACES!!!)
>    export JAVA_HOME="C:/graalvm-ce-java11-22.2.0"
>    Does this mean the location of the Java code for the particular program I want to run with GPU? In my case: C:\StrategyQuantX\j64
>    (folder where I extracted the _graalvm-ee-java17-windows-amd64-22.3.0_ )

TornadoVM for "Windows" is experimental. Indeed, what we currently do is compile TornadoVM through MSys64. I personally use Windows 11 and it seems to work well for PTX and OpenCL. 

Alternatively, you can use WSL2 (e.g., Ubuntu). In my case, it works really well. the installation is exactly the same as any other Linux System, and you can use it also for running on Intel Integrated GPUs, Intel ARC GPUs (discrete) and FPGAs (emulation mode) if you have Intel CPU/GPU. 

To answer your question: No, you do not have to your Java code in the same location as your Java home. You can use `--classpath`, and all Java flags as any other Java program. `tornado` is just an alias for `java <all_tornadovm_paramers> <tornado_modules>` . 

For reference, this is my `source.sh` file for Windows 11:

```bash
#!/bin/bash

export JAVA_HOME="C:\Users\jjfum\Documents\bin\jvms\graalvm-ce-java17-windows-amd64-22.2.0\graalvm-ce-java17-22.2.0"

export CUDA_PATH="C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.6"
export PTX_LDFLAGS=-L\"$CUDA_PATH/lib/x64\"

# LEAVE THE REST OF FILE 'AS IS'
# DON'T ALTER!
export PATH=$PWD/bin/bin:$PATH               ## This directory will be automatically generated during Tornado compilation
export TORNADO_SDK=$PWD/bin/sdk              ## This directory will be automatically generated during Tornado compilation
CMAKE_FILE=$(where cmake | head -n 1)
export CMAKE_ROOT=${CMAKE_FILE%\\*\\*}
```

> 3. Lastly, when I try to compile the program, I get these errors, not sure how to fix them: [Error on building Tornado.txt](https://github.com/beehive-lab/TornadoVM/files/9978042/Error.on.building.Tornado.txt)

It seems there is an error with some of the maven dependencies. Can you try again? 

Also, note that TornadoVM assumes a correct installation of CUDA, or any supported vendor implementation for OpenCL (e.g., Intel, AMD, etc). For example, if you have an NVIDIA GPU, quite likely the driver is already installed. You only need to install the CUDA SDK.
https://developer.nvidia.com/cuda-downloads 

> 
> Thank you!

Hope this helps. Feel free to ask and open new issues. We are keen to get feedback from the community.

**Comment by Excelnabb** [comment]
Hi
Thank you for the answers. 
I tried running the compiling two times, still the same errors. 
[Error on building Tornado second try.txt](https://github.com/beehive-lab/TornadoVM/files/9980590/Error.on.building.Tornado.second.try.txt)

**Reply by jjfumero** [reply]
Can you use Graal CE 22.2.0 instead? Also, can you compile with one of the backends?

```bash
make BACKEND=opencl
```

**Comment by Excelnabb** [comment]
Thank you so much for the help, that worked. I manage to install it. I installed Graal 22.2.0 into the J64 folder, as that was okay as I understod?

However when I run the program, it uses the CPU and not the GPU.
Is there any parameter or code in this folder I need to change to force the program to run on the GPU?
![location](https://user-images.githubusercontent.com/42059629/201103995-97d58d09-904a-4dd1-a827-87072d749a54.jpg)

**Reply by jjfumero** [reply]
The GPU selection should be automatic with TornadoVM. Maybe you have also support for OpenCL CPU?
One way to detect all devices available is via `tornado --devices`. 

You can change device from the command line using 

`--jvm="-D<yourTaskGraphName>.<yourTaskName>.device=<backendIndex><deviceIndex>` 

We have a repo with examples about how to change device, print the generated code, etc. 

https://github.com/jjfumero/tornadovm-examples 

This can help.
